src/services/pcn-bridge/src/Bridge_dp_no_stp.c

```diff
@@ -0,0 +1,310 @@
+} __attribute__((packed));
+
+struct fwd_entry {
+  u64 timestamp;
+  u32 port;
+  enum entry_type type;
+} __attribute__((packed));
+
+BPF_TABLE("hash", struct fwd_key, struct fwd_entry, fwdtable, 1024);
+
+/*
+ * Port Table: Saves the state of the ports
+ * TODO: what is the best option in this case, hash vs array?
+ */
+struct port {
+  u16 mode;
+  u16 native_vlan;
+  bool native_vlan_enabled;
+};
+
+BPF_TABLE("hash", u32, struct port, ports, 256);
+
+struct eth_hdr {
+  __be64 dst : 48;
+  __be64 src : 48;
+  __be16 proto;
+} __attribute__((packed));
+
+static int handle_rx(struct CTXTYPE *ctx, struct pkt_metadata *md) {
+  void *data = (void *)(long)ctx->data;
+  void *data_end = (void *)(long)ctx->data_end;
+  struct eth_hdr *eth = data;
+
+  if (data + sizeof(*eth) > data_end)
+    goto DROP;
+
+  u32 in_ifc = md->in_port;
+
+  pcn_log(ctx, LOG_DEBUG, "in_ifc=%d proto=0x%x", in_ifc,
+          bpf_ntohs(eth->proto));
+  pcn_log(ctx, LOG_DEBUG, "mac src:%M dst:%M", eth->src, eth->dst);
+
+  u16 vlanid = 0;
+#ifdef POLYCUBE_XDP
+  bool tagged = pcn_is_vlan_present(ctx);
+
+  if (tagged) {
+    u16 eth_proto;
+    if (pcn_get_vlan_id(ctx, &vlanid, &eth_proto) < 0)
+      goto DROP;
+  }
+#else
+  bool tagged = ctx->vlan_present;
+
+  if (tagged)
+    vlanid = ctx->vlan_tci & 0x0fff;
+#endif
+
+  struct port *in_port = ports.lookup(&in_ifc);
+  if (!in_port) {
+    pcn_log(ctx, LOG_ERR, "Port %u not found", in_ifc);
+    goto DROP;
+  }
+
+  struct port_vlan_key vlan_key;
+  struct port_vlan_value *vlan_entry;
+
+  vlan_key.port = in_ifc;
+
+  if (in_port->mode == PORT_MODE_ACCESS) {
+    if (tagged)
+      goto DROP;
+
+    vlan_key.vlan = VLAN_WILDCARD;
+    vlan_entry = port_vlan.lookup(&vlan_key);
+    if (!vlan_entry)
+      goto DROP;  // TODO: is this code correct?
+
+    vlanid = vlan_entry->vlan;
+
+    pcn_log(ctx, LOG_TRACE, "port mode ACCESS vlan=%d, state=%d", vlanid,
+            vlan_entry->stp_state);
+  } else {
+    // TRUNK
+    if (!tagged) {
+      vlanid = in_port->native_vlan;
+    }
+    vlan_key.vlan = vlanid;
+    vlan_entry = port_vlan.lookup(&vlan_key);
+    if (!vlan_entry) {
+      // TODO manage case of all vlans allowed
+      goto DROP;
+    }
+  }
+
+  // LEARNING PHASE
+  struct fwd_key src_key;
+  src_key.vlan = vlanid;
+  src_key.mac = eth->src;
+  struct fwd_entry e;  // used below
+  e.timestamp = bpf_ktime_get_ns();
+  e.port = in_ifc;
+  e.type = DYNAMIC;
+
+  // make sure that not exist a static entry associated to this address & vlan
+  struct fwd_entry *src_entry = fwdtable.lookup(&src_key);
+  if (!src_entry || src_entry->type == DYNAMIC) {
+    fwdtable.update(&src_key, &e);
+  }
+
+  // FORWARDING PHASE: select interface(s) to send the packet
+  struct fwd_key dst_key;
+  dst_key.vlan = vlanid;
+  dst_key.mac = eth->dst;
+
+  // lookup in forwarding table fwdtable
+  struct fwd_entry *entry = fwdtable.lookup(&dst_key);
+  if (!entry) {
+    // bpf_trace_printk("entry not found\n");
```

Comment:
Never ever use printk, even commented. We have logging, use that for everything.